File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-reanimated-4-use-worklets-scheduler.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { resolveReanimatedApiName } from "./utils/resolve-reanimated-api-name.js";

const WORKLETS_MIGRATION_BY_REANIMATED_API = new Map<string, string>([
  ["runOnUI", "replace `runOnUI(fn)(...args)` with `scheduleOnUI(fn, ...args)`"],
  ["runOnJS", "replace `runOnJS(fn)(...args)` with `scheduleOnRN(fn, ...args)`"],
  [
    "executeOnUIRuntimeSync",
    "replace `executeOnUIRuntimeSync(fn)(...args)` with `runOnUISync(fn, ...args)`",
  ],
  [
    "runOnRuntime",
    "replace `runOnRuntime(runtime, fn)(...args)` with `scheduleOnRuntime(runtime, fn, ...args)`",
  ],
]);

const REANIMATED_SCHEDULER_API_NAMES: ReadonlySet<string> = new Set(
  WORKLETS_MIGRATION_BY_REANIMATED_API.keys(),
);

export const rnReanimated4UseWorkletsScheduler = defineRule({
  id: "rn-reanimated-4-use-worklets-scheduler",
  title: "Scheduler moved to Worklets",
  tags: ["migration-hint"],
  requires: ["reanimated:4"],
  severity: "warn",
  recommendation:
    "Import the corresponding scheduler from `react-native-worklets` when migrating to Reanimated 4.",
  create: (context: RuleContext) => ({
    CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
      const apiName = resolveReanimatedApiName(
        node,
        context.scopes,
        REANIMATED_SCHEDULER_API_NAMES,
      );
      if (!apiName) return;
      const migration = WORKLETS_MIGRATION_BY_REANIMATED_API.get(apiName);
      if (!migration) return;
      context.report({
        node,
        message: `For Reanimated 4, ${migration} from \`react-native-worklets\`.`,
      });
    },
  }),
});

File: native/oxlint/rules/rn-reanimated-4-use-worklets-scheduler.rs
use oxc_ast::AstKind;
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;

use crate::{AstNode, context::LintContext, rule::Rule};

const REANIMATED_MODULE_SOURCES: [&str; 1] = ["react-native-reanimated"];
const SCHEDULER_MIGRATIONS: [(&str, &str); 4] = [
    (
        "runOnUI",
        "replace `runOnUI(fn)(...args)` with `scheduleOnUI(fn, ...args)`",
    ),
    (
        "runOnJS",
        "replace `runOnJS(fn)(...args)` with `scheduleOnRN(fn, ...args)`",
    ),
    (
        "executeOnUIRuntimeSync",
        "replace `executeOnUIRuntimeSync(fn)(...args)` with `runOnUISync(fn, ...args)`",
    ),
    (
        "runOnRuntime",
        "replace `runOnRuntime(runtime, fn)(...args)` with `scheduleOnRuntime(runtime, fn, ...args)`",
    ),
];

#[derive(Debug, Default, Clone)]
#[allow(non_camel_case_types)]
pub struct RnReanimated_4UseWorkletsScheduler;

pub type RnReanimated4UseWorkletsScheduler = RnReanimated_4UseWorkletsScheduler;

declare_oxc_lint!(
    /// Prefer Reanimated 4 Worklets scheduler APIs.
    RnReanimated_4UseWorkletsScheduler,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Prefer Reanimated 4 Worklets scheduler APIs.",
);

impl Rule for RnReanimated_4UseWorkletsScheduler {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::CallExpression(call_expression) = node.kind() else {
            return;
        };
        let Some((_, migration)) = SCHEDULER_MIGRATIONS.iter().find(|(api_name, _)| {
            module_api_path_matches(
                &call_expression.callee,
                &[*api_name],
                &REANIMATED_MODULE_SOURCES,
                true,
                ctx,
            )
        }) else {
            return;
        };
        ctx.diagnostic(
            OxcDiagnostic::warn(format!(
                "For Reanimated 4, {migration} from `react-native-worklets`."
            ))
            .with_label(call_expression.span),
        );
    }
}
